Make middle click not mousedown close tabs

Zachiah opened the pull request:
Closes #44856

Release Notes:

- Make a full middle click be required to close tabs not just a middle mousedown

Zachiah committed:
Make middle click not mousedown close tabs

Zachiah commented on crates/gpui/src/interactive.rs:
If you would prefer me to inline it, just give me the word. I figured this was useful since the analogous for left/middle already exist.

Zachiah committed:
Add on_right_click and on_left_click events

Zachiah commented:
This PR needs extensive testing. I hope I took the right approach. This is by far my most substantial PR to zed so far so please have grace 🙃. I wonder if the methods I added should be used in other places? I added `on_right_click` even though I didn't need it because it felt crazy to add `on_middle_click` but not `on_right_click`

Zachiah committed:
Underscore unused variables

Zachiah committed:
Remove unused method I added

Zachiah commented:
I added `on_right_click` and `on_middle_click` instead of adding a first parameter to match the API of say `on_mouse_down` because I didn't want to make this a breaking change to gpui and have to update a billion places. If it is preferable to make a breaking change and add it as a first parameter just let me know and I will get working on that. My thought is that if we ever did want to do that we could change it to that from what it is currently fairly easily when we had other breaking changes say. I honestly don't know enough about how gpui is used and if it's fine to break it willy nilly to make the right call on my own.

Anthony-Eid commented on crates/gpui/src/elements/div.rs:
```diff
@@ -522,6 +522,38 @@ impl Interactivity {
+            .push(Rc::new(move |event, window, cx| {
+                listener(event, window, cx)
+            }));
+    }
```
We typically follow MDM standards which would use have `on_aux_click` instead of these two callbacks.
https://developer.mozilla.org/en-US/docs/Web/API/Element/auxclick_event

Anthony-Eid commented:
https://github.com/zed-industries/zed/blob/a628d5d467dda8b9f8f6b1c9234d4a7a877dad6d/crates/gpui/src/elements/div.rs#L2270 You should add the new click listeners you added to this check

Zachiah committed:
Change API; Fix bug

Zachiah commented:
@Anthony-Eid Thanks for the helpful feedback and review. I updated it. Sorry about the force push. The sqlite db had conflicts for me locally so I rebased and that fixed it.

Anthony-Eid committed:
merge with main

Anthony-Eid committed:
Fix is_middle_click doc comment

Anthony-Eid committed:
Polish
